lib/samtools/samtools.1
'\" t
.TH samtools 1 "22 April 2016" "samtools-1.3.1" "Bioinformatics tools"
.SH NAME
samtools \- Utilities for the Sequence Alignment/Map (SAM) format
.\"
.\" Copyright (C) 2008-2011, 2013-2016 Genome Research Ltd.
.\" Portions copyright (C) 2010, 2011 Broad Institute.
.\"
.\" Author: Heng Li <lh3@sanger.ac.uk>
.\" Author: Joshua C. Randall <jcrandall@alum.mit.edu>
.\"
.\" Permission is hereby granted, free of charge, to any person obtaining a
.\" copy of this software and associated documentation files (the "Software"),
.\" to deal in the Software without restriction, including without limitation
.\" the rights to use, copy, modify, merge, publish, distribute, sublicense,
.\" and/or sell copies of the Software, and to permit persons to whom the
.\" Software is furnished to do so, subject to the following conditions:
.\"
.\" The above copyright notice and this permission notice shall be included in
.\" all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\" IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
.\" THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
.\" LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
.\" FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
.\" DEALINGS IN THE SOFTWARE.
.
.\" For code blocks and examples (cf groff's Ultrix-specific man macros)
.de EX

.  in +\\$1
.  nf
.  ft CR
..
.de EE
.  ft
.  fi
.  in

..
.
.SH SYNOPSIS
.PP
samtools view -bt ref_list.txt -o aln.bam aln.sam.gz
.PP
samtools sort -T /tmp/aln.sorted -o aln.sorted.bam aln.bam
.PP
samtools index aln.sorted.bam
.PP
samtools idxstats aln.sorted.bam
.PP
samtools flagstat aln.sorted.bam
.PP
samtools stats aln.sorted.bam
.PP
samtools bedcov aln.sorted.bam
.PP
samtools depth aln.sorted.bam
.PP
samtools view aln.sorted.bam chr2:20,100,000-20,200,000
.PP
samtools merge out.bam in1.bam in2.bam in3.bam
.PP
samtools faidx ref.fasta
.PP
samtools tview aln.sorted.bam ref.fasta
.PP
samtools split merged.bam
.PP
samtools quickcheck in1.bam in2.cram
.PP
samtools dict -a GRCh38 -s "Homo sapiens" ref.fasta
.PP
samtools fixmate in.namesorted.sam out.bam
.PP
samtools mpileup -C50 -gf ref.fasta -r chr3:1,000-2,000 in1.bam in2.bam
.PP
samtools flags PAIRED,UNMAP,MUNMAP
.PP
samtools fastq input.bam > output.fastq
.PP
samtools fasta input.bam > output.fasta
.PP
samtools addreplacerg -r 'ID:fish' -r 'LB:1334' -r 'SM:alpha' -o output.bam input.bam
.PP
samtools collate aln.sorted.bam aln.name_collated.bam
.PP
samtools depad input.bam

.SH DESCRIPTION
.PP
Samtools is a set of utilities that manipulate alignments in the BAM
format. It imports from and exports to the SAM (Sequence Alignment/Map)
format, does sorting, merging and indexing, and allows to retrieve reads
in any regions swiftly.

Samtools is designed to work on a stream. It regards an input file `-'
as the standard input (stdin) and an output file `-' as the standard
output (stdout). Several commands can thus be combined with Unix
pipes. Samtools always output warning and error messages to the standard
error output (stderr).

Samtools is also able to open a BAM (not SAM) file on a remote FTP or
HTTP server if the BAM file name starts with `ftp://' or `http://'.
Samtools checks the current working directory for the index file and
will download the index upon absence. Samtools does not retrieve the
entire alignment file unless it is asked to do so.

.SH COMMANDS AND OPTIONS

.TP 10 \"-------- view
.B view
samtools view
.RI [ options ]
.IR in.sam | in.bam | in.cram
.RI [ region ...]

With no options or regions specified, prints all alignments in the specified
input alignment file (in SAM, BAM, or CRAM format) to standard output
in SAM format (with no header).

You may specify one or more space-separated region specifications after the
input filename to restrict output to only those alignments which overlap the
specified region(s). Use of region specifications requires a coordinate-sorted
and indexed input file (in BAM or CRAM format).

The
.BR -b ,
.BR -C ,
.BR -1 ,
.BR -u ,
.BR -h ,
.BR -H ,
and
.B -c
options change the output format from the default of headerless SAM, and the
.B -o
and
.B -U
options set the output file name(s).

The
.B -t
and
.B -T
options provide additional reference data. One of these two options is required
when SAM input does not contain @SQ headers, and the
.B -T
option is required whenever writing CRAM output.

The
.BR -L ,
.BR -r ,
.BR -R ,
.BR -q ,
.BR -l ,
.BR -m ,
.BR -f ,
and
.B -F
options filter the alignments that will be included in the output to only those
alignments that match certain criteria.

The
.BR -x ,
.BR -B ,
and
.B -s
options modify the data which is contained in each alignment.

Finally, the
.B -@
option can be used to allocate additional threads to be used for compression, and the
.B -?
option requests a long help message.

.TP
.B REGIONS:
.RS
Regions can be specified as: RNAME[:STARTPOS[-ENDPOS]] and all position
coordinates are 1-based.

Important note: when multiple regions are given, some alignments may be output
multiple times if they overlap more than one of the specified regions.

Examples of region specifications:
.TP 10
.B chr1
Output all alignments mapped to the reference sequence named `chr1' (i.e. @SQ SN:chr1).
.TP
.B chr2:1000000
The region on chr2 beginning at base position 1,000,000 and ending at the
end of the chromosome.
.TP
.B chr3:1000-2000
The 1001bp region on chr3 beginning at base position 1,000 and ending at base
position 2,000 (including both end positions).
.TP
.B '*'
Output the unmapped reads at the end of the file.
(This does not include any unmapped reads placed on a reference sequence
alongside their mapped mates.)
.TP
.B .
Output all alignments.
(Mostly unnecessary as not specifying a region at all has the same effect.)
.RE

.B OPTIONS:
.RS
.TP 10
.B -b
Output in the BAM format.
.TP
.B -C
Output in the CRAM format (requires -T).
.TP
.B -1
Enable fast BAM compression (implies -b).
.TP
.B -u
Output uncompressed BAM. This option saves time spent on
compression/decompression and is thus preferred when the output is piped
to another samtools command.
.TP
.B -h
Include the header in the output.
.TP
.B -H
Output the header only.
.TP
.B -c
Instead of printing the alignments, only count them and print the
total number. All filter options, such as
.BR -f ,
.BR -F ,
and
.BR -q ,
are taken into account.
.TP
.B -?
Output long help and exit immediately.
.TP
.BI "-o " FILE
Output to
.I FILE [stdout].
.TP
.BI "-U " FILE
Write alignments that are
.I not
selected by the various filter options to
.IR FILE .
When this option is used, all alignments (or all alignments intersecting the
.I regions
specified) are written to either the output file or this file, but never both.
.TP
.BI "-t " FILE
A tab-delimited
.IR FILE .
Each line must contain the reference name in the first column and the length of
the reference in the second column, with one line for each distinct reference.
Any additional fields beyond the second column are ignored. This file also
defines the order of the reference sequences in sorting. If you run:
`samtools faidx <ref.fa>', the resulting index file
.I <ref.fa>.fai
can be used as this
.IR FILE .
.TP
.BI "-T " FILE
A FASTA format reference
.IR FILE ,
optionally compressed by
.B bgzip
and ideally indexed by
.B samtools
.BR faidx .
If an index is not present, one will be generated for you.
.TP
.BI "-L " FILE
Only output alignments overlapping the input BED
.I FILE
[null].
.TP
.BI "-r " STR
Only output alignments in read group
.I STR
[null].
.TP
.BI "-R " FILE
Output alignments in read groups listed in
.I FILE
[null].
.TP
.BI "-q " INT
Skip alignments with MAPQ smaller than
.I INT
[0].
.TP
.BI "-l " STR
Only output alignments in library
.I STR
[null].
.TP
.BI "-m " INT
Only output alignments with number of CIGAR bases consuming query
sequence \(>=
.I INT
[0]
.TP
.BI "-f " INT
Only output alignments with all bits set in
.I INT
present in the FLAG field.
.I INT
can be specified in hex by beginning with `0x' (i.e. /^0x[0-9A-F]+/)
or in octal by beginning with `0' (i.e. /^0[0-7]+/) [0].
.TP
.BI "-F " INT
Do not output alignments with any bits set in
.I INT
present in the FLAG field.
.I INT
can be specified in hex by beginning with `0x' (i.e. /^0x[0-9A-F]+/)
or in octal by beginning with `0' (i.e. /^0[0-7]+/) [0].
.TP
.BI "-x " STR
Read tag to exclude from output (repeatable) [null]
.TP
.B -B
Collapse the backward CIGAR operation.
.TP
.BI "-s " FLOAT
Integer part is used to seed the random number generator [0]. Part after the
decimal point sets the fraction of templates/pairs to subsample [no subsampling].
.TP
.BI "-@ " INT
Number of BAM compression threads to use in addition to main thread [0].
.TP
.B -S
Ignored for compatibility with previous samtools versions.
Previously this option was required if input was in SAM format, but now the
correct format is automatically detected by examining the first few characters
of input.
.RE

.TP \"-------- sort
.B sort
.na
samtools sort
.RB [ -l
.IR level ]
.RB [ -m
.IR maxMem ]
.RB [ -o
.IR out.bam ]
.RB [ -O
.IR format ]
.RB [ -n ]
.RB [ -T
.IR tmpprefix ]
.RB [ -@
.IR threads "] [" in.sam | in.bam | in.cram ]
.ad

Sort alignments by leftmost coordinates, or by read name when
.B -n
is used.
An appropriate
.B @HD-SO
sort order header tag will be added or an existing one updated if necessary.

The sorted output is written to standard output by default, or to the
specified file
.RI ( out.bam )
when
.B -o
is used.
This command will also create temporary files
.IB tmpprefix . %d .bam
as needed when the entire alignment data cannot fit into memory
(as controlled via the
.B -m
option).

.B Options:
.RS
.TP 11
.BI "-l " INT
Set the desired compression level for the final output file, ranging from 0
(uncompressed) or 1 (fastest but minimal compression) to 9 (best compression
but slowest to write), similarly to
.BR gzip (1)'s
compression level setting.
.IP
If
.B -l
is not used, the default compression level will apply.
.TP
.BI "-m " INT
Approximately the maximum required memory per thread, specified either in bytes
or with a
.BR K ", " M ", or " G
suffix.
[768 MiB]
.TP
.B -n
Sort by read names (i.e., the
.B QNAME
field) rather than by chromosomal coordinates.
.TP
.BI "-o " FILE
Write the final sorted output to
.IR FILE ,
rather than to standard output.
.TP
.BI "-O " FORMAT
Write the final output as
.BR sam ", " bam ", or " cram .

By default, samtools tries to select a format based on the
.B -o
filename extension; if output is to standard output or no format can be
deduced,
.B bam
is selected.
.TP
.BI "-T " PREFIX
Write temporary files to
.IB PREFIX . nnnn .bam,
or if the specified
.I PREFIX
is an existing directory, to
.IB PREFIX /samtools. mmm . mmm .tmp. nnnn .bam,
where
.I mmm
is unique to this invocation of the
.B sort
command.
.IP
By default, any temporary files are written alongside the output file, as
.IB out.bam .tmp. nnnn .bam,
or if output is to standard output, in the current directory as
.BI samtools. mmm . mmm .tmp. nnnn .bam.
.TP
.BI "-@ " INT
Set number of sorting and compression threads.
By default, operation is single-threaded.
.PP
Historically
.B samtools sort
also accepted a less flexible way of specifying the final and
temporary output filenames:
.IP
samtools sort
.RB [ -f "] [" -o ]
.I in.bam out.prefix
.PP
This has now been removed.
The previous \fIout.prefix\fP argument (and \fB-f\fP option, if any)
should be changed to an appropriate combination of \fB-T\fP \fIPREFIX\fP
and \fB-o\fP \fIFILE\fP.  The previous \fB-o\fP option should be removed,
as output defaults to standard output.
.RE

.TP \"-------- index
.B index
samtools index
.RB [ -bc ]
.RB [ -m
.IR INT ]
.IR aln.bam | aln.cram
.RI [ out.index ]

Index a coordinate-sorted BAM or CRAM file for fast random access.
(Note that this does not work with SAM files even if they are bgzip
compressed \(em to index such files, use tabix(1) instead.)

This index is needed when
.I region
arguments are used to limit
.B samtools view
and similar commands to particular regions of interest.

If an output filename is given, the index file will be written to
.IR out.index .
Otherwise, for a CRAM file
.IR aln.cram ,
index file
.IB aln.cram .crai
will be created; for a BAM file
.IR aln.bam ,
either
.IB aln.bam .bai
or
.IB aln.bam .csi
will be created, depending on the index format selected.

.B Options:
.RS
.TP 8
.B -b
Create a BAI index.
This is currently the default when no format options are used.
.TP
.B -c
Create a CSI index.
By default, the minimum interval size for the index is 2^14, which is the same
as the fixed value used by the BAI format.
.TP
.BI "-m " INT
Create a CSI index, with a minimum interval size of 2^INT.
.RE

.TP \"-------- idxstats
.B idxstats
samtools idxstats
.IR in.sam | in.bam | in.cram

Retrieve and print stats in the index file corresponding to the input file.
Before calling idxstats, the input BAM file must be indexed by samtools index.

The output is TAB-delimited with each line consisting of reference sequence
name, sequence length, # mapped reads and # unmapped reads. It is written to
stdout.

.TP \"-------- flagstat
.B flagstat
samtools flagstat
.IR in.sam | in.bam | in.cram

Does a full pass through the input file to calculate and print statistics
to stdout.

Provides counts for each of 13 categories based primarily on bit flags in
the FLAG field. Each category in the output is broken down into QC pass and
QC fail, which is presented as "#PASS + #FAIL" followed by a description of
the category.

The first row of output gives the total number of reads that are QC pass and
fail (according to flag bit 0x200). For example:

  122 + 28 in total (QC-passed reads + QC-failed reads)

Which would indicate that there are a total of 150 reads in the input file,
122 of which are marked as QC pass and 28 of which are marked as "not passing
quality controls"

Following this, additional categories are given for reads which are:

.RS 18
.TP
secondary
0x100 bit set
.TP
supplementary
0x800 bit set
.TP
duplicates
0x400 bit set
.TP
mapped
0x4 bit not set
.TP
paired in sequencing
0x1 bit set
.TP
read1
both 0x1 and 0x40 bits set
.TP
read2
both 0x1 and 0x80 bits set
.TP
properly paired
both 0x1 and 0x2 bits set and 0x4 bit not set
.TP
with itself and mate mapped
0x1 bit set and neither 0x4 nor 0x8 bits set
.TP
singletons
both 0x1 and 0x8 bits set and bit 0x4 not set
.RE

.RS 10
And finally, two rows are given that additionally filter on the reference
name (RNAME), mate reference name (MRNM), and mapping quality (MAPQ) fields:
.RE

.RS 18
.TP
with mate mapped to a different chr
0x1 bit set and neither 0x4 nor 0x8 bits set and MRNM not equal to RNAME
.TP
with mate mapped to a different chr (mapQ>=5)
0x1 bit set and neither 0x4 nor 0x8 bits set
and MRNM not equal to RNAME and MAPQ >= 5
.RE

.TP \"-------- stats
.B stats
samtools stats
.RI [ options ]
.IR in.sam | in.bam | in.cram
.RI [ region ...]

samtools stats collects statistics from BAM files and outputs in a text format.
The output can be visualized graphically using plot-bamstats.

.B Options:
.RS
.TP 8
.BI "-c, --coverage " MIN , MAX , STEP
Set coverage distribution to the specified range (MIN, MAX, STEP all given as integers)
[1,1000,1]
.TP
.B -d, --remove-dups
Exclude from statistics reads marked as duplicates
.TP
.BI "-f, --required-flag "  STR "|" INT
Required flag, 0 for unset. See also `samtools flags`
[0]
.TP
.BI "-F, --filtering-flag " STR "|" INT
Filtering flag, 0 for unset. See also `samtools flags`
[0]
.TP
.BI "--GC-depth " FLOAT
the size of GC-depth bins (decreasing bin size increases memory requirement)
[2e4]
.TP
.B -h, --help
This help message
.TP
.BI "-i, --insert-size " INT
Maximum insert size
[8000]
.TP
.BI "-I, --id " STR
Include only listed read group or sample name
[]
.TP
.BI "-l, --read-length " INT
Include in the statistics only reads with the given read length
[]
.TP
.BI "-m, --most-inserts " FLOAT
Report only the main part of inserts
[0.99]
.TP
.BI "-P, --split-prefix " STR
A path or string prefix to prepend to filenames output when creating
categorised statistics files with
.BR -S / --split .
[input filename]
.TP
.BI "-q, --trim-quality " INT
The BWA trimming parameter
[0]
.TP
.BI "-r, --ref-seq " FILE
Reference sequence (required for GC-depth and mismatches-per-cycle calculation).
[]
.TP
.BI "-S, --split " TAG
In addition to the complete statistics, also output categorised statistics
based on the tagged field
.I TAG
(e.g., use
.B --split RG
to split into read groups).

Categorised statistics are written to files named
.RI < prefix >_< value >.bamstat,
where
.I prefix
is as given by
.B --split-prefix
(or the input filename by default) and
.I value
has been encountered as the specified tagged field's value in one or more
alignment records.
.TP
.BI "-t, --target-regions " FILE
Do stats in these regions only. Tab-delimited file chr,from,to, 1-based, inclusive.
[]
.TP
.B "-x, --sparse"
Suppress outputting IS rows where there are no insertions.
.RE

.TP \"-------- bedcov
.B bedcov
samtools bedcov
.IR region.bed " " in1.sam | in1.bam | in1.cram "[...]"

Reports read depth per genomic region, as specified in the supplied BED file.

.TP \"-------- depth
.B depth
samtools depth
.RI [ options ]
.RI "[" in1.sam | in1.bam | in1.cram " [" in2.sam | in2.bam | in2.cram "] [...]]"

Computes the depth at each position or region.

.B Options:
.RS
.TP 8
.B -a
Output all positions (including those with zero depth)
.TP
.B -a -a, -aa
Output absolutely all positions, including unused reference sequences
.TP
.BI "-b "  FILE
.RI "Compute depth at list of positions or regions in specified BED " FILE.
[]
.TP
.BI "-f " FILE
.RI "Use the BAM files specified in the " FILE
(a file of filenames, one file per line)
[]
.TP
.BI "-l " INT
.RI "Ignore reads shorter than " INT
.TP
.BI "-m, -d " INT
.RI "Truncate reported depth at a maximum of " INT " reads."
[8000]
.TP
.BI "-q " INT
.RI "Only count reads with base quality greater than " INT
.TP
.BI "-Q " INT
.RI "Only count reads with mapping quality greater than " INT
.TP
.BI "-r " CHR ":" FROM "-" TO
Only report depth in specified region.
.RE

.TP \"-------- merge
.B merge
samtools merge [-nur1f] [-h inh.sam] [-R reg] [-b <list>] <out.bam> <in1.bam> [<in2.bam> <in3.bam> ... <inN.bam>]

Merge multiple sorted alignment files, producing a single sorted output file
that contains all the input records and maintains the existing sort order.

If
.BR -h
is specified the @SQ headers of input files will be merged into the specified header, otherwise they will be merged
into a composite header created from the input headers.  If in the process of merging @SQ lines for coordinate sorted
input files, a conflict arises as to the order (for example input1.bam has @SQ for a,b,c and input2.bam has b,a,c)
then the resulting output file will need to be re-sorted back into coordinate order.

Unless the
.BR -c
or
.BR -p
flags are specified then when merging @RG and @PG records into the output header then any IDs found to be duplicates
of existing IDs in the output header will have a suffix appended to them to differentiate them from similar header
records from other files and the read records will be updated to reflect this.

.B OPTIONS:
.RS
.TP 8
.B -1
Use zlib compression level 1 to compress the output.
.TP
.BI -b \ FILE
List of input BAM files, one file per line.
.TP
.B -f
Force to overwrite the output file if present.
.TP 8
.BI -h \ FILE
Use the lines of
.I FILE
as `@' headers to be copied to
.IR out.bam ,
replacing any header lines that would otherwise be copied from
.IR in1.bam .
.RI ( FILE
is actually in SAM format, though any alignment records it may contain
are ignored.)
.TP
.B -n
The input alignments are sorted by read names rather than by chromosomal
coordinates
.TP
.BI -R \ STR
Merge files in the specified region indicated by
.I STR
[null]
.TP
.B -r
Attach an RG tag to each alignment. The tag value is inferred from file names.
.TP
.B -u
Uncompressed BAM output
.TP
.B -c
When several input files contain @RG headers with the same ID, emit only one
of them (namely, the header line from the first file we find that ID in) to
the merged output file.
Combining these similar headers is usually the right thing to do when the
files being merged originated from the same file.

Without \fB-c\fP, all @RG headers appear in the output file, with random
suffices added to their IDs where necessary to differentiate them.
.TP
.B -p
Similarly, for each @PG ID in the set of files to merge, use the @PG line
of the first file we find that ID in rather than adding a suffix to
differentiate similar IDs.
.RE

.TP \"-------- faidx
.B faidx
samtools faidx <ref.fasta> [region1 [...]]

Index reference sequence in the FASTA format or extract subsequence from
indexed reference sequence. If no region is specified,
.B faidx
will index the file and create
.I <ref.fasta>.fai
on the disk. If regions are specified, the subsequences will be
retrieved and printed to stdout in the FASTA format.

The input file can be compressed in the
.B BGZF
format.

The sequences in the input file should all have different names.
If they do not, indexing will emit a warning about duplicate sequences and
retrieval will only produce subsequences from the first sequence with the
duplicated name.

.TP \"-------- tview
.B tview
samtools tview
.RB [ -p
.IR chr:pos ]
.RB [ -s
.IR STR ]
.RB [ -d
.IR display ]
.RI <in.sorted.bam>
.RI [ref.fasta]

Text alignment viewer (based on the ncurses library). In the viewer,
press `?' for help and press `g' to check the alignment start from a
region in the format like `chr10:10,000,000' or `=10,000,000' when
viewing the same reference sequence.

.B Options:
.RS
.TP 14
.BI -d \ display
Output as (H)tml or (C)urses or (T)ext
.TP
.BI -p \ chr:pos
Go directly to this position
.TP
.BI -s \ STR
Display only alignments from this sample or read group
.RE

.TP \"-------- split
.B split
samtools split
.RI [ options ]
.IR merged.sam | merged.bam | merged.cram

Splits a file by read group.

.B Options:
.RS
.TP 14
.BI "-u " FILE1
.RI "Put reads with no RG tag or an unrecognised RG tag into " FILE1
.TP
.BI "-u " FILE1 ":" FILE2
.RI "As above, but assigns an RG tag as given in the header of " FILE2
.TP
.BI "-f " STRING
Output filename format string (see below)
["%*_%#.%."]
.TP
.B -v
Verbose output
.PP
Format string expansions:
.TS
center;
lb l .
%%	%
%*	basename
%#	@RG index
%!	@RG ID
%.	output format filename extension
.TE
.RE

.TP \"-------- quickcheck
.B quickcheck
samtools quickcheck
.RI [ options ]
.IR in.sam | in.bam | in.cram
[ ... ]

Quickly check that input files appear to be intact. Checks that beginning of the
file contains a valid header (all formats) containing at least one target
sequence and then seeks to the end of the file and checks that an end-of-file
(EOF) is present and intact (BAM only).

Data in the middle of the file is not read since that would be much more time
consuming, so please note that this command will not detect internal corruption,
but is useful for testing that files are not truncated before performing more
intensive tasks on them.

This command will exit with a non-zero exit code if any input files don't have a
valid header or are missing an EOF block. Otherwise it will exit successfully
(with a zero exit code).

.B Options:
.RS
.TP 8
.B -v
Verbose output: will additionally print the names of all input files that don't
pass the check to stdout. Multiple -v options will cause additional messages
regarding check results to be printed to stderr.
.RE

.TP \"-------- dict
.B dict
samtools dict <ref.fasta|ref.fasta.gz>

Create a sequence dictionary file from a fasta file.

.B OPTIONS:
.RS
.TP 11
.BI -a,\ --assembly \ STR
Specify the assembly for the AS tag.
.TP
.B -H,\ --no-header
Do not print the @HD header line.
.TP
.BI -o,\ --output \ FILE
Output to
.I FILE
[stdout].
.TP
.BI -s,\ --species \ STR
Specify the species for the SP tag.
.TP
.BI -u,\ --uri \ STR
Specify the URI for the UR tag. Defaults to
the absolute path of
.I ref.fasta
unless reading from stdin.
.RE

.TP \"-------- fixmate
.B fixmate
.na
samtools fixmate
.RB [ -rpc ]
.RB [ -O
.IR format ]
.I in.nameSrt.bam out.bam
.ad

Fill in mate coordinates, ISIZE and mate related flags from a
name-sorted alignment.

.B OPTIONS:
.RS
.TP 11
.B -r
Remove secondary and unmapped reads.
.TP
.B -p
Disable FR proper pair check.
.TP
.B -c
Add template cigar ct tag.
.TP
.BI "-O " FORMAT
Write the final output as
.BR sam ", " bam ", or " cram .

By default, samtools tries to select a format based on the output
filename extension; if output is to standard output or no format can be
deduced,
.B bam
is selected.
.RE

.TP \"-------- mpileup
.B mpileup
samtools mpileup
.RB [ -EBugp ]
.RB [ -C
.IR capQcoef ]
.RB [ -r
.IR reg ]
.RB [ -f
.IR in.fa ]
.RB [ -l
.IR list ]
.RB [ -Q
.IR minBaseQ ]
.RB [ -q
.IR minMapQ ]
.I in.bam
.RI [ in2.bam
.RI [ ... ]]

Generate VCF, BCF or pileup for one or multiple BAM files. Alignment records
are grouped by sample (SM) identifiers in @RG header lines. If sample
identifiers are absent, each input file is regarded as one sample.

In the pileup format (without
.BR -u \ or \ -g ),
each
line represents a genomic position, consisting of chromosome name,
1-based coordinate, reference base, the number of reads covering the site,
read bases, base qualities and alignment
mapping qualities. Information on match, mismatch, indel, strand,
mapping quality and start and end of a read are all encoded at the read
base column. At this column, a dot stands for a match to the reference
base on the forward strand, a comma for a match on the reverse strand,
a '>' or '<' for a reference skip, `ACGTN' for a mismatch on the forward
strand and `acgtn' for a mismatch on the reverse strand. A pattern
`\\+[0-9]+[ACGTNacgtn]+' indicates there is an insertion between this
reference position and the next reference position. The length of the
insertion is given by the integer in the pattern, followed by the
inserted sequence. Similarly, a pattern `-[0-9]+[ACGTNacgtn]+'
represents a deletion from the reference. The deleted bases will be
presented as `*' in the following lines. Also at the read base column, a
symbol `^' marks the start of a read. The ASCII of the character
following `^' minus 33 gives the mapping quality. A symbol `$' marks the
end of a read segment.

Note that there are two orthogonal ways to specify locations in the
input file; via \fB-r\fR \fIregion\fR and \fB-l\fR \fIfile\fR.  The
former uses (and requires) an index to do random access while the
latter streams through the file contents filtering out the specified
regions, requiring no index.  The two may be used in conjunction.  For
example a BED file containing locations of genes in chromosome 20
could be specified using \fB-r 20 -l chr20.bed\fR, meaning that the
index is used to find chromosome 20 and then it is filtered for the
regions listed in the bed file.

.B Input Options:
.RS
.TP 10
.B -6, --illumina1.3+
Assume the quality is in the Illumina 1.3+ encoding.
.TP
.B -A, --count-orphans
Do not skip anomalous read pairs in variant calling.
.TP
.BI -b,\ --bam-list \ FILE
List of input BAM files, one file per line [null]
.TP
.B -B, --no-BAQ
Disable probabilistic realignment for the computation of base alignment
quality (BAQ). BAQ is the Phred-scaled probability of a read base being
misaligned. Applying this option greatly helps to reduce false SNPs
caused by misalignments.
.TP
.BI -C,\ --adjust-MQ \ INT
Coefficient for downgrading mapping quality for reads containing
excessive mismatches. Given a read with a phred-scaled probability q of
being generated from the mapped position, the new mapping quality is
about sqrt((INT-q)/INT)*INT. A zero value disables this
functionality; if enabled, the recommended value for BWA is 50. [0]
.TP
.BI -d,\ --max-depth \ INT
At a position, read maximally
.I INT
reads per input file. Note that samtools has a minimum value of
.I 8000/n
where
.I n
is the number of input files given to mpileup.  This means the default
is highly likely to be increased.  Once above the cross-sample minimum of
8000 the -d parameter will have an effect. [250]
.TP
.B -E, --redo-BAQ
Recalculate BAQ on the fly, ignore existing BQ tags
.TP
.BI -f,\ --fasta-ref \ FILE
The
.BR faidx -indexed
reference file in the FASTA format. The file can be optionally compressed by
.BR bgzip .
[null]
.TP
.BI -G,\ --exclude-RG \ FILE
Exclude reads from readgroups listed in FILE (one @RG-ID per line)
.TP
.BI -l,\ --positions \ FILE
BED or position list file containing a list of regions or sites where
pileup or BCF should be generated. Position list files contain two
columns (chromosome and position) and start counting from 1.  BED
files contain at least 3 columns (chromosome, start and end position)
and are 0-based half-open.
.br
While it is possible to mix both position-list and BED coordinates in

pyamd/samtools.py
import os
import sys
import time
import logging
import argparse
import subprocess

class Samtools:

    def __init__(self, sam_path, bft_path, out_path):
        self.sam_path = sam_path
        self.out_path = out_path
        self.bft_path = bft_path


    def fixmate(self, bam_path):
        base = os.path.splitext(os.path.basename(bam_path))[0]
        obam_path = '{0}/{1}_fixmate.bam'.format(self.out_path, base)
        fmcmd = [self.sam_path, 'fixmate', '-O', 'BAM',
                bam_path, obam_path]
        fmrun = subprocess.Popen(fmcmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=False)
        fmrun.wait()
#        if fmrun.returncode != 0:
#            logger.error('Samtools fixmate failed running the following command : {0}'.format(' '.join(fmcmd)))

        return(obam_path, fmrun.returncode)


    def sort(self, bam_path):
        base = os.path.splitext(os.path.basename(bam_path))[0]
        obam_path = '{0}/{1}_sorted.bam'.format(self.out_path, base)
        stcmd = [self.sam_path, 'sort', '-O', 'BAM',
                '-o', obam_path, bam_path]
        strun = subprocess.Popen(stcmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=False)
        strun.wait()
#        if strun.returncode != 0:
#            logger.error('Samtools sort failed running the following command : {0}'.format(' '.join(stcmd)))

        return(obam_path, strun.returncode)

    def addreadgroup(self, bam_path, sam_name):
        run_time = time.strftime('%d/%m/%Y')
        base = os.path.splitext(os.path.basename(bam_path))[0]
        abam_path = '{0}/{1}_RG.bam'.format(self.out_path, base)
        acmd = [self.sam_path, 'addreplacerg', '-O', 'BAM',
                '-r', '"@RG\tID:{0}\tPG:{0}\tSM:{0}\tPM:{0}\tLB=MaRS\tPL=Illumina\tPU=MiSeq\tDT={1}\tPI=null"'.format(sam_name, run_time),
                '-o', abam_path, bam_path]
        arun = subprocess.Popen(acmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=False)
        arun.wait()
        icmd = [self.sam_path, 'index', abam_path]
        irun = subprocess.Popen(icmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=False)
        irun.wait()
        if arun.returncode != 0:
            print(' '.join(acmd))
        return(abam_path, arun.returncode)

    def pileup(self, ref_path, bam_path):
        obcf_path = '{0}/variants.bcf'.format(self.out_path)

        mpcmd = [self.sam_path, 'mpileup', '-go', obcf_path,
                '-f', ref_path, bam_path]
        mprun = subprocess.Popen(mpcmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=False)
        mprun.wait()
#        if mprun.returncode != 0:
#            logger.error('Samtools mpileup failed running the following command : {0}'.format(' '.join(mpcmd)))

        return(obcf_path, mprun.returncode)

    def bcfindex(self, bcf_path):

        bicmd = [self.bft_path, 'index', bcf_path]
        birun = subprocess.Popen(bicmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=False)
        birun.wait()
#        if birun.returncode != 0:
#            logger.error('Bcftools index failed running the following command : {0}'.format(' '.join(bicmd)))

        return(birun.returncode)

    def bcftools(self, bcf_path, bed_path, sam_name):
        ovcf_path = '{0}/{1}_variants.vcf'.format(self.out_path, sam_name)
        btcmd = [self.bft_path, 'call', '--skip-variants', 'indels',
                '--multiallelic-caller', '--variants-only', '-O', 'v',
                '-s', sam_name, '-o', ovcf_path, bcf_path]
        btrun = subprocess.Popen(btcmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=False)
        btrun.wait()
#        if btrun.returncode != 0:
#            logger.error('Bcftools call failed running the following command : {0}'.format(' '.join(btcmd)))

        return(ovcf_path, btrun.returncode)


    def bcfstats(self, vcf_path, ref_path):
        stats_path = '{0}/variants.stats'.format(self.out_path)
        bscmd = [self.bft_path, 'stats', '-F', ref_path, '-s' , '-',
                vcf_path, '>', stats_path]
        bsrun = subprocess.Popen(' '.join(bscmd), stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True)
        bsrun.wait()
#        if bsrun.returncode != 0:
#            logger.error('Bcftools stats failed running the following command : {0}'.format(' '.join(bscmd)))

        return(stats_path, bsrun.returncode)
